server/gateway.go

```diff
@@ -1459,31 +1645,41 @@ func (c *client) processGatewaySubjectSub(arg []byte) error {
 // given account/subject (which means, it does not have a registered
 // no-interest on the account and/or subject) and the sublist result
 // for queue subscriptions.
+// <Outbound connection: invoked when client message is published,
+// so from any client connection's readLoop>
 func (c *client) gatewayInterest(acc, subj string) (bool, *SublistResult) {
-	sni, accountInMap := c.gw.noInterest.Load(acc)
-	// If there is an entry for this account and sni is nil,
+	ei, accountInMap := c.gw.outsim.Load(acc)
+	// If there is an entry for this account and ei is nil,
 	// it means that the remote is not interested at all in
 	// this account and we could not possibly have queue subs.
-	if accountInMap && sni == nil {
+	if accountInMap && ei == nil {
 		return false, nil
 	}
+	// Assume interest if account not in map.
+	psi := !accountInMap
 	var r *SublistResult
-	// Get the matches for queue subs.
-	sli, _ := c.gw.qsubsInterest.Load(acc)
-	if sli != nil {
-		sl := sli.(*Sublist)
-		r = sl.Match(subj)
-	}
-	// Represents plain sub interest
-	psi := true
 	if accountInMap {
-		// If there is a value, sni is itself a sync.Map for subject
-		// no-interest (in that account).
-		subjectNoInterest := sni.(*sync.Map)
-		if _, subjInMap := subjectNoInterest.Load(subj); subjInMap {
-			// If subject is present in map, it means a no-interest
-			// for this subject.
-			psi = false
+		// If in map, check for subs interest with sublist.
+		e := ei.(*outsie)
+		r = e.sl.Match(subj)
+		// If there is plain subs returned, we don't have to
+		// check if we should use the no-interest map because
+		// it means that we are in perfect knowledge mode.
+		// Only if there is nothing returned for r.psubs that
+		// we need to check.
+		if len(r.psubs) > 0 {
+			psi = true
+		} else {
+			e.RLock()
+			// We may be in transition to perfect knowledge mode
+			// (e.sas is true), but until e.ni is nil, use it to
```

Comment:
maybe make sas more clear that we have transitioned to a perfect knowledge mode? Maybe mode on gateway with enums of OPTIMISTIC and PERFECT or something like that?

Reply:
seems like a good idea.